inventory.lua

```diff
@@ -83,63 +102,109 @@ minetest.register_node("digilines:chest", {
+		-- Remember what was in the target slot before the put; see
+		-- on_metadata_inventory_put for why we care.
+		last_inventory_put_listname = listname
+		last_inventory_put_index = index
+		last_inventory_put_stack = minetest.get_meta(pos):get_inventory():get_stack(listname, index)
 		return stack:get_count()
 	end,
-	on_metadata_inventory_move = function(pos, _, _, _, _, _, player)
+	allow_metadata_inventory_take = function(_, listname, index, stack)
+		-- Remember the listname and index values; see tube.remove_items for
+		-- why we care.
+		last_inventory_take_listname = listname
+		last_inventory_take_index = index
+		return stack:get_count()
+	end,
+	on_metadata_inventory_move = function(pos, from_list, from_index, to_list, to_index, count, player)
 		minetest.log("action", player:get_player_name().." moves stuff in chest at "..minetest.pos_to_string(pos))
 	end,
-	on_metadata_inventory_put = function(pos, _, _, stack, player)
-		local channel = minetest.get_meta(pos):get_string("channel")
-		local send = function(msg)
-			sendMessage(pos,msg,channel)
-		end
-		-- direction is only for furnaces
-		-- as the item has already been put, can_insert should return false if the chest is now full.
-		local derpstack = stack:get_name()..' 1'
-		if can_insert(pos,derpstack) then
-			send("uput "..maybeString(stack))
+	on_metadata_inventory_put = function(pos, listname, index, stack, player)
+		-- Get what was in the target slot before the put; it has disappeared
+		-- by now (been replaced by the result of the put action) but we saved
+		-- it in allow_metadata_inventory_put. This should always work
+		-- (allow_metadata_inventory_put should AFAICT always be called
+		-- immediately before on_metadata_inventory_put), but in case of
+		-- something weird happening, just fall back to using an empty
+		-- ItemStack rather than crashing.
+		local old_stack
+		if last_inventory_put_listname == listname and last_inventory_put_index == index then
+			old_stack = last_inventory_put_stack
+			last_inventory_put_listname = nil
+			last_inventory_put_index = nil
+			last_inventory_put_stack = nil
 		else
-			send("ufull "..maybeString(stack))
+			old_stack = ItemStack(nil)
+		end
+		-- If the player tries to place a stack into an inventory, there’s
+		-- already a stack there, and the existing stack is either of a
+		-- different item type or full, then obviously the stacks can’t be
+		-- merged; instead the stacks are swapped. This information is not
+		-- reported to mods (Minetest core neither tells us that a particular
+		-- action was a swap, nor tells us a take followed by a put). In core,
+		-- the condition for swapping is that you try to add the new stack to
+		-- the existing stack and the leftovers are as big as the original
+		-- stack to put. Replicate that logic here using the old stack saved in
+		-- allow_metadata_inventory_put. If a swap happened, report it to the
+		-- Digilines network as a utake followed by a uput.
```

Comment:
Oh hell...
You’d better [submit an issue](https://github.com/minetest/minetest/issues/new).

Reply:
What exactly should I say? Should swaps be reported in a fresh callback which is different to all the existing ones? Or should a swap report as a take followed by a put? I guess if the latter that wouldn’t break compatibility with anything because anyone who replicated the logic I wrote would just never see that happen (the take would happen first, leaving the put now able to put things and the leftovers would thus not be as big as the original any more). But should an `allow_metadata_inventory_take` also happen before permitting a swap? I guess probably?

Reply:
@Hawk777 Let the developers decide.

Reply:
minetest/minetest#6534